File: python/sol/input_data.py
# -*- coding: utf-8 -*-
# input_data.py

import numpy as np

def read(fn, Parm):

    # 計算パラメーター(辞書型)
    Parm['title']      = ''
    Parm['solver']     = [1.95, 1000, 50, 1e-5]
    Parm['plot3dgeom'] = 0

    # 変数初期化
    #ierr = 0
    Nx = Ny = Nz = 0
    Xn = Yn = Zn = None

    # ファイルを読み込む
    with open(fn, encoding='utf-8') as fp:
        header = fp.readline()
        lines = fp.readlines()

    # 配列の大きさを調べる
    # 誘電体の電圧と電極の誘電率のために+1しておく
    nvolt = 1
    nepsr = 1
    ngeometry = 0
    for tline in lines:
        if tline.strip().startswith('volt = '):
            nvolt += 1
        if tline.strip().startswith('epsr = '):
            nepsr += 1
        if tline.strip().startswith('geometry = '):
            ngeometry += 1
    if ngeometry < 2:
        print('*** number of geometry data < 2')
        return

    # 配列を確保する
    fVolt = np.zeros(nvolt, 'f8')
    fEpsr = np.zeros(nepsr, 'f8')
    iGeometry = np.zeros((ngeometry, 2), 'i4')
    fGeometry = np.zeros((ngeometry, 8), 'f8')
    
    # 誘電体の電圧と電極の比誘電率
    fVolt[0] = 0
    fEpsr[0] = 1   # 静電エネルギー計算のために必要

    # ヘッダー
    d = header.strip().split()
    if (len(d) < 3) or (d[0] != 'OpenSTF'):
        print('*** Not OpenSTF data : %s' % fn)
        return
    version = (100 * int(d[1])) + int(d[2])
    if version < 301:
        print('*** version of data is old : %s.%s < 3.1' % (d[1], d[2]))
        return

    # 本体
    nvolt = 1
    nepsr = 1
    ngeometry = 0
    for tline in lines:
        # token
        d = tline.strip().split()  #;print(d)
        
        # check
        if d[0].startswith('#'):
            continue
        elif (d[0] == 'end'):
            break
        elif (len(d) < 3):
            continue
        elif (d[1] != "="):
            continue

        # parameter array
        p = d[2:]  #;print(p)
        
        # input
        key = d[0]
        if   key == 'title':
            Parm['title'] = tline[len('title = '):].strip()
        elif key == 'xmesh':
            Nx, Xn = _makemesh(p)
        elif key == 'ymesh':
            Ny, Yn = _makemesh(p)
        elif key == 'zmesh':
            Nz, Zn = _makemesh(p)
        elif key == 'volt':
            fVolt[nvolt] = float(p[0])
            nvolt += 1
        elif key == 'epsr':
            fEpsr[nepsr] = float(p[0])
            nepsr += 1
        elif key == 'geometry':
            if len(p) > 8:
                # p[0]=1/2 : 電極/誘電体
                # p[1]=1,2,3,... : 属性番号
                index = int(p[1])
                shape = int(p[2])
                # 電極のとき: ID=-1,-2,-3,...(属性番号に負号をつける)
                # 誘電体のとき; ID=1,2,3,...
                iGeometry[ngeometry, 0] = -index if p[0] == '1' else index
                # 形状番号=1,2,11,...
                iGeometry[ngeometry, 1] = shape
                # 座標値(6個または8個)
                #ndata = 6 if shape in [1, 2, 11, 12, 13] else 8
                ndata = 0
                if shape in [1, 2, 11, 12, 13]:
                    ndata = 6
                elif shape in [31, 32, 33, 41, 42, 43, 51, 52, 53]:
                    ndata = 8
                fGeometry[ngeometry, 0:ndata] = [float(g) for g in p[3:3+ndata]]
                ngeometry += 1
        elif key == 'solver':
            if len(p) > 3:
                Parm['solver'] = [float(p[0]), int(p[1]), int(p[2]), float(p[3])]
        elif key == 'plot3dgeom':
            Parm['plot3dgeom'] = int(p[0])  # new

    # error check

    if Nx <= 0:
        print('*** no X mesh')
        #ierr = 1
    if Ny <= 0:
        print('*** no Y mesh')
        #ierr = 1
    if Nz <= 0:
        print('*** no Z mesh')
        #ierr = 1

    if nvolt < 1 + 2:
        # 電圧は最低2個必要
        print('*** number of volt.s < 2')
        #ierr = 1

    for n in range(iGeometry.shape[0]):
        if (iGeometry[n, 0] < 0) and (-iGeometry[n, 0] > nvolt):
            print('*** invalid geometry (volt) index')
            #ierr = 1
        if (iGeometry[n, 0] > 0) and (+iGeometry[n, 0] > nepsr):
            print('*** invalid geometry (epsr) index')
            #ierr = 1
        if not iGeometry[n, 1] in [1, 2, 11, 12, 13, 31, 32, 33, 41, 42, 43, 51, 52, 53]:
            print('*** invalid geometry index')
            #ierr = 1

    if (Parm['solver'][1] <= 0) or (Parm['solver'][2] <= 0):
        print('*** invalid solver data')
        #ierr = 1

    # 厚さのない物体形状を節点に寄せる
    _fit_geometry(Xn, Yn, Zn, iGeometry, fGeometry)

    #print(Parm)
    #print(Nx, Ny, Nz)
    #print(Xn, Yn, Zn)
    #print(fVolt)
    #print(fEpsr)
    #print(iGeometry)
    #print(fGeometry)
    return Parm, Nx, Ny, Nz, Xn, Yn, Zn, fVolt, fEpsr, iGeometry, fGeometry

# (private) 節点データ作成
def _makemesh(p):
    EPS = 1e-12

    # error check
    if len(p) < 3:  # データ数は最低3個
        print('*** number of mesh data < 3')
        return None
    if (len(p) % 2 == 1) and (float(p[0]) > float(p[-1])):
        p = p[::-1]  # 座標が降順なら昇順に変換する

    # 区間配列作成
    pos = np.zeros(0, 'f8')
    div = np.zeros(0, 'i4')
    for i, d in enumerate(p):
        if i % 2 == 0:
            # 区間座標
            if (i > 0) and (abs(float(d) - pos[-1]) < EPS):
                # 区間幅が0ならその区間を削除する
                div = np.delete(div, i // 2 - 1)
                continue
            pos = np.append(pos, float(d))
        else:
            # 区間分割数
            if int(d) <= 0:
                # 区間分割数は正
                print('*** division of mesh <= 0')
                return None
            div = np.append(div, int(d))

    # error check
    if (len(pos) < 2) or (len(div) < 1):
        # 区間座標は最低2個、区間分割数は最低1個必要
        print('*** no valid mesh data')
        return None

    # 節点データ作成
    node = np.array(pos[0], float)  # 最初
    for i in range(len(div)):
        dl = (pos[i + 1] - pos[i]) / div[i]
        node = np.append(node, np.linspace(pos[i] + dl, pos[i + 1], div[i]))
    #print(node)
    
    return len(node) - 1, node

# 厚さのない物体形状を節点に寄せる
def _fit_geometry(Xn, Yn, Zn, iGeometry, fGeometry):

    if (Xn is None) or (Yn is None) or (Zn is None):
        return

    # 長さ次元の微小量
    eps = 1e-6 * (
        abs(Xn[0] - Xn[-1]) +
        abs(Yn[0] - Yn[-1]) +
        abs(Zn[0] - Zn[-1]))

    for n in range(iGeometry.shape[0]):
        s = iGeometry[n, 1]
        p = fGeometry[n, :]
        x1, x2, y1, y2, z1, z2 = p[0:6]

        if   (s == 1) or (s == 2):
            # rectangle, ellipsoid
            if abs(x1 - x2) < eps:
                i = np.argmin(abs(Xn - x1))
                p[0:2] = Xn[i]
            if abs(y1 - y2) < eps:
                j = np.argmin(abs(Yn - y1))
                p[2:4] = Yn[j]
            if abs(z1 - z2) < eps:
                k = np.argmin(abs(Zn - z1))
                p[4:6] = Zn[k]
        elif s == 11:
            # X cylinder
            if abs(x1 - x2) < eps:
                i = np.argmin(abs(Xn - x1))
                p[0:2] = Xn[i]
        elif s == 12:
            # Y cylinder
            if abs(y1 - y2) < eps:
                j = np.argmin(abs(Yn - y1))
                p[2:4] = Yn[j]
        elif s == 13:
            # Z cylinder
            if abs(z1 - z2) < eps:
                k = np.argmin(abs(Zn - z1))
                p[4:6] = Zn[k]

File: sol/input_data.c
/*
input_data.c

input data
*/

#include "ost.h"
#include "ost_prototype.h"

static void setup_cells(int, int, int, int *, int *, int *);
static void setup_node(int, int, int, double *, double *, double *, int *, int *, int *);
static void fitgeometry(void);
static void setup_geomlines(void);

#define MAXTOKEN 1000

int input_data(FILE *fp)
{
	int    ntoken, ngeom, nline;
	//int    version = 0;
	int    nxr = 0, nyr = 0, nzr = 0;
	int    *dxr = NULL, *dyr = NULL, *dzr = NULL;
	double *xr = NULL, *yr = NULL, *zr = NULL;
	char   strline[BUFSIZ], strkey[BUFSIZ], strsave[BUFSIZ];
	char   *token[MAXTOKEN];
	const int array_inc = 10000;		// reduce malloc times
	const char sep[] = " \t";			// separator

	// initialize

	NVolt = 1;
	NEpsr = 1;
	Volt[0] = 0;
	Epsr[0] = 1;

	NGeom = 0;
	Geom = NULL;

	Solver.omega = 1.95;
	Solver.maxiter = 1000;
	Solver.nout = 50;
	Solver.converg = 1e-5;

	Plot3dGeom = 0;

	// read

	nline = 0;
	while (fgets(strline, sizeof(strline), fp) != NULL) {
		//printf("%s", strline);

		// skip a vacant line
		if (strlen(strline) <= 1) continue;

		// skip a comment line
		if (strline[0] == '#') continue;

		// delete "\n"
		//printf("%ld\n", strlen(strline));
		if (strstr(strline, "\r\n") != NULL) {
			strline[strlen(strline) - 2] = '\0';
		}
		else if ((strstr(strline, "\r") != NULL) || (strstr(strline, "\n") != NULL)) {
			strline[strlen(strline) - 1] = '\0';
		}
		//printf("%ld\n", strlen(strline));

		// "end" -> break
		if (!strncmp(strline, "end", 3)) break;

		// save "strline"
		strcpy(strsave, strline);

		// token ("strline" is destroyed)
		ntoken = tokenize(strline, sep, token, MAXTOKEN);
		//for (int i = 0; i < ntoken; i++) printf("%d %s\n", i, token[i]);

		// check number of data and "=" (excluding header)
		if ((nline > 0) && ((ntoken < 3) || strcmp(token[1], "="))) continue;

		// keyword
		strcpy(strkey, token[0]);

		// input
		if      (nline == 0) {
			if (strcmp(strkey, "OpenSTF")) {
				fprintf(stderr, "*** not OpenSTF data\n");
				return 1;
			}
			else if (ntoken > 2) {
				//version = (10 * atoi(token[1])) + atoi(token[2]);
			}
			nline++;
		}
		else if (!strcmp(strkey, "title")) {
			strcpy(Title, strchr(strsave, '=') + 2);
			//strncpy(Title, strchr(strsave, '=') + 2, sizeof(Title));
		}
		else if (!strcmp(strkey, "xmesh")) {
			if ((ntoken < 5) || (ntoken % 2 == 0)) {
				fprintf(stderr, "*** invalid %s data\n", strkey);
				return 1;
			}
			nxr = (ntoken - 3) / 2;
			xr = (double *)malloc((nxr + 1) * sizeof(double));
			dxr = (int *)malloc(nxr * sizeof(int));
			sscanf(token[2], "%lf", &xr[0]);
			for (int i = 0; i < nxr; i++) {
				sscanf(token[2 * i + 3], "%d", &dxr[i]);
				sscanf(token[2 * i + 4], "%lf", &xr[i + 1]);
			}
		}
		else if (!strcmp(strkey, "ymesh")) {
			if ((ntoken < 5) || (ntoken % 2 == 0)) {
				fprintf(stderr, "*** invalid %s data\n", strkey);
				return 1;
			}
			nyr = (ntoken - 3) / 2;
			yr = (double *)malloc((nyr + 1) * sizeof(double));
			dyr = (int *)malloc(nyr * sizeof(int));
			sscanf(token[2], "%lf", &yr[0]);
			for (int j = 0; j < nyr; j++) {
				sscanf(token[2 * j + 3], "%d", &dyr[j]);
				sscanf(token[2 * j + 4], "%lf", &yr[j + 1]);
			}
		}
		else if (!strcmp(strkey, "zmesh")) {
			if ((ntoken < 5) || (ntoken % 2 == 0)) {
				fprintf(stderr, "*** invalid %s data\n", strkey);
				return 1;
			}
			nzr = (ntoken - 3) / 2;
			zr = (double *)malloc((nzr + 1) * sizeof(double));
			dzr = (int *)malloc(nzr * sizeof(int));
			sscanf(token[2], "%lf", &zr[0]);
			for (int k = 0; k < nzr; k++) {
				sscanf(token[2 * k + 3], "%d", &dzr[k]);
				sscanf(token[2 * k + 4], "%lf", &zr[k + 1]);
			}
		}
		else if (!strcmp(strkey, "volt")) {
			if (NVolt >= 256) {
				fprintf(stderr, "*** too many %s data #%d\n", strkey, NVolt + 1);
				return 1;
			}
			Volt[NVolt++] = atof(token[2]);
		}
		else if (!strcmp(strkey, "epsr")) {
			if (NEpsr >= 256) {
				fprintf(stderr, "*** too many %s data #%d\n", strkey, NEpsr + 1);
				return 1;
			}
			Epsr[NEpsr++] = (real_t)atof(token[2]);
		}
		else if (!strcmp(strkey, "geometry")) {
			if (ntoken < 5) {
				fprintf(stderr, "*** invalid %s data #%d\n", strkey, NGeom + 1);
				return 1;
			}
			if (NGeom % array_inc == 0) {
				Geom = (geometry_t *)realloc(Geom, (NGeom + array_inc) * sizeof(geometry_t));
			}
			Geom[NGeom].type  = atoi(token[2]);        // =1/2
			Geom[NGeom].pid   = (id_t)atoi(token[3]);  // =1,2,...
			Geom[NGeom].shape = atoi(token[4]);        // =1,2,...
			switch (Geom[NGeom].shape) {
				case 1:
				case 2:
				case 11:
				case 12:
				case 13:
					ngeom = 6;
					break;
				case 31:
				case 32:
				case 33:
				case 41:
				case 42:
				case 43:
				case 51:
				case 52:
				case 53:
					ngeom = 8;
					break;
				default:
					ngeom = 0;
					break;
			}
			if (ntoken < 5 + ngeom) {
				fprintf(stderr, "*** invalid %s data #%d\n", strkey, NGeom + 1);
				return 1;
			}
			for (int n = 0; n < ngeom; n++) {
				Geom[NGeom].g[n] = atof(token[5 + n]);
			}
			NGeom++;
		}
		else if (!strcmp(strkey, "solver")) {
			if (ntoken > 5) {
				Solver.omega   = atof(token[2]);
				Solver.maxiter = atoi(token[3]);
				Solver.nout    = atoi(token[4]);
				Solver.converg = atof(token[5]);
			}
		}
		else if (!strcmp(strkey, "plot3dgeom")) {
			Plot3dGeom = atoi(token[2]);
		}
	}

	// debug
	//printf("title = %s\n", Title);
	//printf("xmesh = %e", xr[0]); for (int i = 0; i < nxr; i++) printf(" %d %e", dxr[i], xr[i + 1]); printf("\n");
	//printf("ymesh = %e", yr[0]); for (int j = 0; j < nyr; j++) printf(" %d %e", dyr[j], yr[j + 1]); printf("\n");
	//printf("zmesh = %e", zr[0]); for (int k = 0; k < nzr; k++) printf(" %d %e", dzr[k], zr[k + 1]); printf("\n");
	//for (int n = 0; n < NVolt; n++) printf("volt = %d %f\n", n, Volt[n]);
	//for (int n = 0; n < NEpsr; n++) printf("epsr = %d %f\n", n, Epsr[n]);
	//for (int n = 0; n < NGeom; n++) printf("geometry = %d %d %d %d %f %f %f %f %f %f\n", n, Geom[n].type, Geom[n].pid, Geom[n].shape, Geom[n].g[0], Geom[n].g[1], Geom[n].g[2], Geom[n].g[3], Geom[n].g[4], Geom[n].g[5]);
	//printf("solver = %f %d %d %e\n", Solver.omega, Solver.maxiter, Solver.nout, Solver.converg);

	// error check

	if (nxr <= 0) {
		fprintf(stderr, "*** no xmesh data\n");
		return 1;
	}
	if (nyr <= 0) {
		fprintf(stderr, "*** no ymesh data\n");
		return 1;
	}
	if (nzr <= 0) {
		fprintf(stderr, "*** no zmesh data\n");
		return 1;
	}
	for (int i = 0; i < nxr; i++) {
		if ((xr[i] > xr[i + 1]) || (dxr[i] <= 0)) {
			fprintf(stderr, "*** invalid xmesh data\n");
			return 1;
		}
	}
	for (int j = 0; j < nyr; j++) {
		if ((yr[j] > yr[j + 1]) || (dyr[j] <= 0)) {
			fprintf(stderr, "*** invalid ymesh data\n");
			return 1;
		}
	}
	for (int k = 0; k < nzr; k++) {
		if ((zr[k] > zr[k + 1]) || (dzr[k] <= 0)) {
			fprintf(stderr, "*** invalid zmesh data\n");
			return 1;
		}
	}
	if (NVolt < 3) {
		fprintf(stderr, "*** number of volt data < 2\n");
		return 1;
	}
	if ((Solver.maxiter <= 0) || (Solver.nout <= 0)) {
		fprintf(stderr, "*** invalid solver data\n");
		return 1;
	}

	for (int n = 0; n < NEpsr; n++) {
		if (Epsr[n] <= 0) {
			fprintf(stderr, "*** invalid epsr data #%d\n", n + 1);
			return 1;
		}
	}

	for (int n = 0; n < NGeom; n++) {
		if ((Geom[n].type < 1) || (Geom[n].type > 2)) {
			fprintf(stderr, "*** invalid type of geometry data #%d\n", n + 1);
			return 1;
		}
		if (Geom[n].pid < 1) {
			fprintf(stderr, "*** invalid property of geometry data #%d\n", n + 1);
			return 1;
		}
		if (Geom[n].shape < 1) {
			fprintf(stderr, "*** invalid shape of geometry data #%d\n", n + 1);
			return 1;
		}
		if ((Geom[n].type == 1) && (Geom[n].pid > NVolt)) {
			fprintf(stderr, "*** invalid volt of geometry data #%d\n", n + 1);
			return 1;
		}
		if ((Geom[n].type == 2) && (Geom[n].pid > NEpsr)) {
			fprintf(stderr, "*** invalid epsr of geometry data #%d\n", n + 1);
			return 1;
		}
	}

	// number of cells
	setup_cells(nxr, nyr, nzr, dxr, dyr, dzr);

	// node
	setup_node(nxr, nyr, nzr, xr, yr, zr, dxr, dyr, dzr);

	// fit geometry without thickness
	fitgeometry();

	// geometry lines
	setup_geomlines();

	// free
	free(xr);
	free(yr);
	free(zr);
	free(dxr);
	free(dyr);
	free(dzr);

	return 0;
}


// number of cells
static void setup_cells(int nxr, int nyr, int nzr, int *dxr, int *dyr, int *dzr)
{
	int xsum = 0;
	for (int i = 0; i < nxr; i++) {
		xsum += dxr[i];
	}
	Nx = xsum;

	int ysum = 0;
	for (int j = 0; j < nyr; j++) {
		ysum += dyr[j];
	}
	Ny = ysum;

	int zsum = 0;
	for (int k = 0; k < nzr; k++) {
		zsum += dzr[k];
	}
	Nz = zsum;
}

// node
static void setup_node(int nxr, int nyr, int nzr, double *xr, double *yr, double *zr, int *dxr, int *dyr, int *dzr)
{
	if ((nxr <= 0) || (nyr <= 0) || (nzr <= 0)) return;

	Xn = (double *)malloc((Nx + 1) * sizeof(double));
	Yn = (double *)malloc((Ny + 1) * sizeof(double));
	Zn = (double *)malloc((Nz + 1) * sizeof(double));

	int xid = 0;
	for (int mx = 0; mx < nxr; mx++) {
		double dx = (xr[mx + 1] - xr[mx]) / dxr[mx];
		for (int i = 0; i < dxr[mx]; i++) {
			Xn[xid++] = xr[mx] + (i * dx);
		}
	}
	Xn[xid] = xr[nxr];

	int yid = 0;
	for (int my = 0; my < nyr; my++) {
		double dy = (yr[my + 1] - yr[my]) / dyr[my];
		for (int j = 0; j < dyr[my]; j++) {
			Yn[yid++] = yr[my] + (j * dy);
		}
	}
	Yn[yid] = yr[nyr];

	int zid = 0;
	for (int mz = 0; mz < nzr; mz++) {
		double dz = (zr[mz + 1] - zr[mz]) / dzr[mz];
		for (int k = 0; k < dzr[mz]; k++) {
			Zn[zid++] = zr[mz] + (k * dz);
		}
	}
	Zn[zid] = zr[nzr];

	// debug
	//for (int i = 0; i <= Nx; i++) printf("Xn[%d]=%.5f\n", i, Xn[i] * 1e3);
	//for (int j = 0; j <= Ny; j++) printf("Yn[%d]=%.5f\n", j, Yn[j] * 1e3);
	//for (int k = 0; k <= Nz; k++) printf("Zn[%d]=%.5f\n", k, Zn[k] * 1e3);
}


// fit geometry without thickness to the nearest node
static void fitgeometry(void)
{
	if ((Nx <= 0) || (Ny <= 0) || (Nz <= 0)) return;

	double d0 = EPS * (
		fabs(Xn[Nx] - Xn[0]) +
		fabs(Yn[Ny] - Yn[0]) +
		fabs(Zn[Nz] - Zn[0]));

	for (int n = 0; n < NGeom; n++) {
		int    shape = Geom[n].shape;
		double *g = Geom[n].g;

		if      ((shape == 1) || (shape == 2)) {
			if (fabs(g[0] - g[1]) < d0) {
				int i = nearest(g[0], 0, Nx, Xn);
				g[0] = g[1] = Xn[i];
			}
			if (fabs(g[2] - g[3]) < d0) {
				int j = nearest(g[2], 0, Ny, Yn);
				g[2] = g[3] = Yn[j];
			}
			if (fabs(g[4] - g[5]) < d0) {
				int k = nearest(g[4], 0, Nz, Zn);
				g[4] = g[5] = Zn[k];
			}
		}
		else if ((shape == 11) || (shape == 64) || (shape == 65)) {
			if (fabs(g[0] - g[1]) < d0) {
				int i = nearest(g[0], 0, Nx, Xn);
				g[0] = g[1] = Xn[i];
			}
		}
		else if ((shape == 12) || (shape == 61) || (shape == 66)) {
			if (fabs(g[2] - g[3]) < d0) {
				int j = nearest(g[2], 0, Ny, Yn);
				g[2] = g[3] = Yn[j];
			}
		}
		else if ((shape == 13) || (shape == 62) || (shape == 63)) {
			if (fabs(g[4] - g[5]) < d0) {
				int k = nearest(g[4], 0, Nz, Zn);
				g[4] = g[5] = Zn[k];
			}
		}
	}
}


// geometry lines
static void setup_geomlines(void)
{
	// alloc work array
	int *shape     =         (int *)malloc(NGeom * sizeof(int));
	int *id        =         (int *)malloc(NGeom * sizeof(int));
	double (*g)[8] = (double (*)[8])malloc(NGeom * 8 * sizeof(double));

	// set work array
	for (int n = 0; n < NGeom; n++) {
		shape[n] = Geom[n].shape;
		id[n] = Geom[n].type;
		memcpy(g[n], Geom[n].g, 8 * sizeof(double));
	}

	// get array size
	NGline = geomlines(0, NGeom, shape, NULL, NULL, NULL, NULL, 0);

	// alloc data
	Gline  = (double (*)[2][3])malloc(NGline * 2 * 3 * sizeof(double));
	MGline =            (int *)malloc(NGline * sizeof(int));
	memset(Gline, 0, NGline * 2 * 3 * sizeof(double));
	memset(MGline, 0, NGline * sizeof(int));

	// set data
	const double eps = EPS * (fabs(Xn[Nx] - Xn[0]) + fabs(Yn[Ny] - Yn[0]) + fabs(Zn[Nz] - Zn[0]));
	geomlines(1, NGeom, shape, id, g, Gline, MGline, eps);

	// free work array
	free(shape);
	free(id);
	free(g);
}
